### ISSUE TITLE
Hellhound crashes after a short while

### ISSUE THREAD

**Issue by Tatsuya79** [libretro/px68k-libretro#22]
This game works fine in winx68k 0.65 in windows and in px68k stand-alone 0.15 on android.
It crashes 10-15s after starting to play in Retroarch.

While troubleshooting, I tried to disable MIDI with MIDI_SW=false in config.
With other emulators it just makes the game switch from MIDI to internal sound while booting.

Here in Retroarch it goes into an error loop spamming:
BusError: eafa09
func = cpu_readmem24 addr = eafa09 flag = 1

**Comment by Tatsuya79** [comment]
I went with more testing of upstream px68k on Android.
The game freeze after a short while happened in version 0.10 and got fixed with 0.11.

The fix should be into those 2 commits:
https://github.com/hissorii/px68k/commit/f719f1f78c8f116787fa255ce150a976b7a626d6
https://github.com/hissorii/px68k/commit/8e10743bcf4bb42b345471c2de931579c3fc87b6

**Comment by Tatsuya79** [comment]
Those other commits would be interesting to investigate too if they make more sense for you than they do for me.
Perhaps not related with Hellhound, but could be. 
Particularly that bus error looping when disabling midi? (never could reproduce that bug on stand-alone)

https://github.com/hissorii/px68k/commit/76269164a4e6f0cdb9c76c1fa06e0342abde0648
https://github.com/hissorii/px68k/commit/41869748654322157c7f8764aab8ecc5eaf241c0

**Comment by Tatsuya79** [comment]
Is this related to the DC68K_NO_JUMP_TABLE tag in makefile.libretro and the left out m68000/c68k/gen68k.c ?
I see some code similar to what I'm trying to port there but I don't understand it.

**Comment by r-type** [comment]
no gen68k.c is only needed for generate c68k files. when regen file with DC68K_NO_JUMP_TABLE , i 
deleted them for cleanup.
BTW original file could be retrieve here
https://github.com/kenyahiro/px68k/commits/master

**Comment by r-type** [comment]
to build the c68k core i used this cmd.

 gcc gen68k.c -DC68K_GEN -DC68K_NO_JUMP_TABLE -o gencpu

if we want to add the fix in the first commit , I think it's here.

https://github.com/kenyahiro/px68k/blob/master/m68000/c68k/gen68k.inc#L882

**Comment by r-type** [comment]
to tweak 
 wf_op("\tCPU->flag_X = CPU->flag_C = ((src & res & 1) + (src >> 1) + (res >> 1)) >> 23;\n");

**Comment by r-type** [comment]
I've done the mod for (subx/addx), can you test it ?
it doesn’t freeze for me any more, but command for moving are odd.
I would like to see if there is no regression with this mod.
 
http://dl.free.fr/jVTKTV1ZL

**Comment by r-type** [comment]
for the record i change
set_addx_flag
//ORIG	        wf_op("\tCPU->flag_X = CPU->flag_C = ((src & dst & 1) + (src >> 1) + (dst >> 1)) >> 23;\n");
	        wf_op("\tCPU->flag_X = CPU->flag_C = ((((src & 1) + (dst & 1) + (((CPU->flag_X >> 8) & 1) & 1)) >> 1) + (src >> 1) + (dst >> 1)) >> 23;\n");

set_subx_flag
//ORIG	        wf_op("\tCPU->flag_X = CPU->flag_C = ((src & res & 1) + (src >> 1) + (res >> 1)) >> 23;\n");
	        wf_op("\tCPU->flag_X = CPU->flag_C = ((((src & 1) + (res & 1) + (((CPU->flag_X >> 8) & 1) & 1)) >> 1) + (src >> 1) + (res >> 1)) >> 23;\n");

**Comment by r-type** [comment]
internal music works also

**Comment by Tatsuya79** [comment]
Great! Does not freeze any more! 👍 
Didn't notice anything wrong with the controls, have to test more...

Internal music already worked before.
The problem is a bus error loop when changing MIDI_SW to false in config.
It still happens with your changes but retro-wertz told me he managed to make it switch to internal sound, idk how yet.

**Comment by r-type** [comment]
cool ,I will backport the change then .
i will look all to see if there are other fix to backport to new c68k core.
Did the sound works find with ssf2 in android with standalone?
I really would like to find a way to fix under linux 64 without having to do heavy change like changing the size of ptr ...

**Comment by Tatsuya79** [comment]
Just tried sf2ce on stand-alone android, internal music works fine.

I couldn't start the game at first because I had MIDI_SW=false... 
This option is annoying. I'll leave it on true now.

**Comment by r-type** [comment]
leave it open for the MIDI_SW buserr

**Comment by Tatsuya79** [comment]
Just tested max speed with Akazukin Chachacha.
750fps before, 1000fps now. :)

**Comment by retro-wertz** [comment]
this is re-confimed to be working on newly regenerated cpu core.

**Comment by r-type** [comment]
Yes you are right , but i prefere regenreate the core (and fix gencpu.inc) instead of the PR

**Comment by r-type** [comment]
BTW does it mean that only addx fix the helllhound ?

**Comment by retro-wertz** [comment]
it looks like it, but not sure. @Tatsuya79 should be able to answer that,
he was who found that this was the fix for hellhound(and probably other
related bug)

**Comment by r-type** [comment]
originally  hissorii patched for subx and addx . i will keep the two for now. i ill update patch

**Comment by Tatsuya79** [comment]
For Hellhound only that one https://github.com/hissorii/px68k/commit/f719f1f78c8f116787fa255ce150a976b7a626d6 is enough.
Then I added references to some other hissorii did.

**Comment by Tatsuya79** [comment]
Perhaps some fixes here need to be ported as well?

https://github.com/hissorii/px68k/commit/41869748654322157c7f8764aab8ecc5eaf241c0
https://github.com/hissorii/px68k/commit/8e9cbb0038210660c44bea23473a60e8438b4d67

**Comment by retro-wertz** [comment]
we should really try to port the BusErrHandling section of px68k, this is needed at least for x6800 since most games try to detect midi ports, and will revert to internal midi if BusErrHandling was emulated(since we do not have midi port).

i manage to port the code but its not working right. i only managed to get something like this:
![chourensha 68k v1 01 1995 famibe no yosshin -170616-153418](https://user-images.githubusercontent.com/23315382/27225492-08ac9b10-52cd-11e7-9a34-2878de523a9c.png)

code : https://hastebin.com/irevadevid.php
if you read the value from **(PC - CPU->BasePC)) - 2** and  **GET_SR**, it matches the ones in the screenshot, so its probably near correct. im not experience enough with programming, so im not sure where to go from here.

**Comment by r-type** [comment]
Now ,
If i boot with MIDI_SW=false , HellBound. music works fine and i it change the config to  MIDI_SW=true.

**Comment by gingerbeardman** [comment]
@r-type game is 超連射68K (Chō Ren Sha 68K) https://en.wikipedia.org/wiki/Cho_Ren_Sha_68K
